File: python/ffi_wrappers.py
import cbor2
from my_plugin import ffi

def offer(f):
    def offered(input, input_len, output, output_len):
        result = cbor2.dumps(f(*cbor2.loads(bytes(ffi.buffer(input, input_len)))))
        result_ptr = ffi.cast("uint8_t *", ffi.from_buffer(result))
        output[0] = result_ptr
        output_len[0] = len(result)
        DONT_FORGET[result_ptr] += 1
    return offered

def call(f):
    def called(*args):
        args_bin = cbor2.dumps(args)
        result_ptr = ffi.new("uint8_t **")
        result_len = ffi.new("size_t *")
        f(ffi.from_buffer(args_bin), len(args_bin), result_ptr, result_len)
        result = cbor2.loads(bytes(ffi.buffer(result_ptr, result_len)))
        return result
    return called

File: c/ffi_wrappers.c
cbor_item_t *call_cbor(void (*f)(uint8_t *, size_t, uint8_t **, size_t *), cbor_item_t *args)
{
  uint8_t *input;
  size_t input_len, length = cbor_serialize_alloc(args, &input, &input_len);

  uint8_t *res;
  size_t res_len;

  f(input, input_len, &res, &res_len);

  struct cbor_load_result result;

  // TODO: Do something with the result ...
  cbor_item_t *result_cbor = cbor_load(res, res_len, &result);

  free(input);
  free(res);

  return result_cbor;
}
